openscenario/map_fragment/include/map_fragment/geometry.hpp

```diff
@@ -0,0 +1,306 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef MAP_FRAGMENT__GEOMETRY__HPP_
+#define MAP_FRAGMENT__GEOMETRY__HPP_
+
+#include <cmath>
+#include <string>
+#include <algorithm>
+#include <memory>
+#include <stdexcept>
+
+namespace map_fragment
+{
+
+struct Point2d
+{
+  double x = 0;
+  double y = 0;
+};
+
+struct Vector2d
+{
+  double x = 0;
+  double y = 0;
+};
+
+auto operator+(Point2d const& a, Vector2d const& b) -> Point2d
+{
+  return {a.x + b.x, a.y + b.y};
+}
+
+auto operator*(Vector2d const& v, double multiplier) -> Vector2d
+{
+  return {v.x * multiplier, v.y * multiplier};
+}
+
+auto operator-(Point2d const& a, Point2d const& b) -> Vector2d
+{
+  return {a.x - b.x, a.y - b.y};
+}
+
+auto operator+(Vector2d const& a, Vector2d const& b) -> Vector2d
+{
+  return {a.x + b.x, a.y + b.y};
+}
+
+auto rotate(Vector2d const& v, double angle) -> Vector2d
+{
+  return {
+    v.x * std::cos(angle) - v.y * std::sin(angle),
+    v.x * std::sin(angle) + v.y * std::cos(angle)
+  };
+}
+
+auto rotate(Point2d const& p, double angle) -> Point2d
+{
+  Point2d p0 = {0, 0};
+  return p0 + rotate(p - p0, angle);
+}
+
+auto calculateAngle(Vector2d const& v) -> double
+{
+  return std::atan2(v.y, v.x);
+}
+
+struct Transformation2d
+{
+  Vector2d translation;
+  double rotation;
+};  // struct Transformation2d
+
+auto applyTransformation(Point2d const& p, Transformation2d const& t) -> Point2d
+{
+  return rotate(p, t.rotation) + t.translation;
+}
+
+auto applyTransformation(Vector2d const& v, Transformation2d const& t) -> Vector2d
+{
+  return rotate(v, t.rotation);
+}
+
+auto chainTransformations(Transformation2d const& first,
+                          Transformation2d const& second) -> Transformation2d
+{
+  Transformation2d t;
+  t.translation = first.translation + rotate(second.translation, first.rotation);
+  t.rotation = first.rotation + second.rotation;
+  return t;
+}
+
+/**
+ * Parametric representation of a smooth curve in 2D Cartesian space
+ * 
+ * By convention, the curve is assumed to always start at point (0, 0)
+ * and have a tangent vector there of [1, 0]. Geometric trasformations
+ * can be applied afterwards if other start positions and/or
+ * directions are required.
+ */
+class ParametricCurve
+{
+
+public:
+  using Pointer = std::shared_ptr<ParametricCurve>;
+  
+  /**
+   * Calculate curve point for given parameter t ∈ [0, 1]
+   */
+  auto getPosition(double t) -> Point2d
+  {
+    validateCurveParameterOrThrow(t);
+    return getPosition_(t);
+  }
+
+  /**
+   * Calculate curve tangent vector for given parameter t ∈ [0, 1]
+   */
+  auto getTangentVector(double t) -> Vector2d
+  {
+    validateCurveParameterOrThrow(t);
+
+    // TODO Throw error if the vector is not unit?
+    return getTangentVector_(t);
+  }
+
+private:
+  virtual auto getPosition_(double t) -> Point2d = 0;
+  virtual auto getTangentVector_(double t) -> Vector2d = 0;
```

Comment:
Use const for immutable member functions like getters. For example, `auto f(int x) const -> int`.